angular/src/app/datacart/treetable/treetable.component.ts

```diff
@@ -323,31 +331,50 @@ export class TreetableComponent implements OnInit, AfterViewInit {
      * @param event - change event
      */
     cartChanged(which) {
-        console.log("Updating view for change in cart");
+        console.log("Updating view for change in cart contents");
+
+        // update selections due to cart additions or changes to item data
         let node: CartTreeNode = null;
         for (let item of this.dataCart.getFiles()) {
             node = this.dataTree.upsertNodeFor(item);
-            if (item.isSelected && (!this.tt || !this.tt.isSelected(node))) {
+            if (item.isSelected && (!this.tt || !this.tt.isSelected(node))) 
                 this.selectedData.push(node);
-                if (this.tt) this.tt.propagateSelectionUp(node, true);
+            if (this.tt) {
+                this.tt.propagateSelectionDown(node, item.isSelected);
+                this.tt.propagateSelectionUp(node, item.isSelected);
             }
         }
 
-        // remove deleted items
-        this.dataTree.cleanNodes(this.dataCart, (node) => {
-            if (this.tt && this.tt.isSelected(node)) {
-                this.tt.propagateSelectionDown(node, false);
-                this.tt.propagateSelectionUp(node, false);
-            }
-        });
+        // remove deleted items and adjust selection view accordingly
+        this.dataTree.cleanNodes(this.dataCart,
+                                 (node) => {
+                                     if (this.tt && this.tt.isSelected(node))
```

Comment:
The `cleanNodes()` function removes nodes for items that were removed from the cart. This is the trickier part as I need to give it two call back functions that get called each time it removes a node.  The first (`removing`) get called before it is actually removed; this call-back unselects the node.  

The second (`refreshParent`) gets called after a node is removed and takes as it argument the node's parent node.  It's job is to refresh the ancestors selection status.  `propagateSelectionUp()` does this, but it really only works correctly if called on a leaf node; thus, I need to first find a leaf node below the parent (any leaf node will do) to pass to the function.